File: central-services/api/push.ts
import type { VercelRequest, VercelResponse } from '@vercel/node'
import webpush from 'web-push'
import { getHubSupabase, getAuthenticatedUser } from './_lib/supabase.js'

const VAPID_SUBJECT = process.env.VAPID_SUBJECT!
const VAPID_PUBLIC_KEY = process.env.VAPID_PUBLIC_KEY!
const VAPID_PRIVATE_KEY = process.env.VAPID_PRIVATE_KEY!

webpush.setVapidDetails(VAPID_SUBJECT, VAPID_PUBLIC_KEY, VAPID_PRIVATE_KEY)

/**
 * POST /api/push?action=subscribe — register a push subscription (authed user)
 * DELETE /api/push?action=subscribe — unregister a push subscription (authed user)
 * POST /api/push?action=notify — send push notification (server-to-server, service key auth)
 */
export default async function handler(req: VercelRequest, res: VercelResponse) {
  const action = req.query.action as string

  if (action === 'notify' && req.method === 'POST') {
    return handleNotify(req, res)
  }

  if (action === 'subscribe') {
    return handleSubscribe(req, res)
  }

  return res.status(400).json({ error: 'Missing or invalid action query param' })
}

async function handleSubscribe(req: VercelRequest, res: VercelResponse) {
  const user = await getAuthenticatedUser(req, res)
  if (!user) return

  const supabase = getHubSupabase()

  if (req.method === 'POST') {
    const { endpoint, keys } = req.body as {
      endpoint?: string
      keys?: { p256dh?: string; auth?: string }
    }

    if (!endpoint || !keys?.p256dh || !keys?.auth) {
      return res.status(400).json({ error: 'Missing subscription fields' })
    }

    const { error } = await supabase
      .from('push_subscriptions')
      .upsert(
        { user_id: user.id, endpoint, p256dh: keys.p256dh, auth: keys.auth },
        { onConflict: 'user_id,endpoint' }
      )

    if (error) return res.status(500).json({ error: error.message })
    return res.json({ ok: true })
  }

  if (req.method === 'DELETE') {
    const { endpoint } = req.body as { endpoint?: string }
    if (!endpoint) return res.status(400).json({ error: 'Missing endpoint' })

    const { error } = await supabase
      .from('push_subscriptions')
      .delete()
      .eq('user_id', user.id)
      .eq('endpoint', endpoint)

    if (error) return res.status(500).json({ error: error.message })
    return res.json({ ok: true })
  }

  return res.status(405).json({ error: 'Method not allowed' })
}

async function handleNotify(req: VercelRequest, res: VercelResponse) {
  const authHeader = req.headers.authorization
  if (!authHeader?.startsWith('Bearer ')) {
    return res.status(401).json({ error: 'Missing authorization' })
  }

  const token = authHeader.slice(7)
  const serviceKey = process.env.HUB_SUPABASE_SERVICE_ROLE_KEY

  if (token !== serviceKey) {
    const supabase = getHubSupabase()
    const { data: client } = await supabase
      .from('forumline_oauth_clients')
      .select('id')
      .eq('client_secret_hash', token)
      .single()

    if (!client) {
      return res.status(401).json({ error: 'Invalid authorization' })
    }
  }

  const { forumline_id, user_id, title, body, link, forum_domain } = req.body as {
    forumline_id?: string
    user_id?: string
    title: string
    body: string
    link?: string
    forum_domain?: string
  }

  if (!title || !body) {
    return res.status(400).json({ error: 'Missing title or body' })
  }

  const supabase = getHubSupabase()

  let targetUserId = user_id
  if (!targetUserId && forumline_id) {
    const { data: profile } = await supabase
      .from('hub_profiles')
      .select('id')
      .eq('id', forumline_id)
      .single()

    if (!profile) return res.status(404).json({ error: 'User not found' })
    targetUserId = profile.id
  }

  if (!targetUserId) {
    return res.status(400).json({ error: 'Missing user_id or forumline_id' })
  }

  // Check if forum is muted
  if (forum_domain) {
    const { data: forum } = await supabase
      .from('forumline_forums')
      .select('id')
      .eq('domain', forum_domain)
      .single()

    if (forum) {
      const { data: membership } = await supabase
        .from('forumline_memberships')
        .select('notifications_muted')
        .eq('user_id', targetUserId)
        .eq('forum_id', forum.id)
        .single()

      if (membership?.notifications_muted) {
        return res.json({ ok: true, skipped: 'forum_muted' })
      }
    }
  }

  const { data: subscriptions } = await supabase
    .from('push_subscriptions')
    .select('endpoint, p256dh, auth')
    .eq('user_id', targetUserId)

  if (!subscriptions || subscriptions.length === 0) {
    return res.json({ ok: true, sent: 0 })
  }

  const payload = JSON.stringify({ title, body, link, forum_domain })

  let sent = 0
  const staleEndpoints: string[] = []

  for (const sub of subscriptions) {
    try {
      await webpush.sendNotification(
        { endpoint: sub.endpoint, keys: { p256dh: sub.p256dh, auth: sub.auth } },
        payload
      )
      sent++
    } catch (err: any) {
      if (err.statusCode === 410 || err.statusCode === 404) {
        staleEndpoints.push(sub.endpoint)
      } else {
        console.error('[Hub:PushNotify] Failed to send:', err.statusCode, err.body)
      }
    }
  }

  if (staleEndpoints.length > 0) {
    await supabase
      .from('push_subscriptions')
      .delete()
      .eq('user_id', targetUserId)
      .in('endpoint', staleEndpoints)
  }

  return res.json({ ok: true, sent })
}

File: services/forumline-api/handler/push.go
package handler

import (
	"encoding/json"
	"net/http"
	"os"
	"strings"

	"github.com/forumline/forumline/backend/auth"
	"github.com/forumline/forumline/services/forumline-api/service"
	"github.com/forumline/forumline/services/forumline-api/store"
)

type PushHandler struct {
	Store       *store.Store
	PushService *service.PushService
}

func NewPushHandler(s *store.Store, ps *service.PushService) *PushHandler {
	return &PushHandler{Store: s, PushService: ps}
}

func (h *PushHandler) Handle(w http.ResponseWriter, r *http.Request) {
	action := r.URL.Query().Get("action")
	if action == "notify" && r.Method == http.MethodPost {
		h.HandleNotify(w, r)
		return
	}
	if action == "subscribe" {
		h.handleSubscribe(w, r)
		return
	}
	writeJSON(w, http.StatusBadRequest, map[string]string{"error": "Missing or invalid action query param"})
}

func (h *PushHandler) handleSubscribe(w http.ResponseWriter, r *http.Request) {
	userID := auth.UserIDFromContext(r.Context())
	ctx := r.Context()

	if r.Method == http.MethodPost {
		var body struct {
			Endpoint string `json:"endpoint"`
			Keys     struct {
				P256dh string `json:"p256dh"`
				Auth   string `json:"auth"`
			} `json:"keys"`
		}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid request body"})
			return
		}
		if body.Endpoint == "" || body.Keys.P256dh == "" || body.Keys.Auth == "" {
			writeJSON(w, http.StatusBadRequest, map[string]string{"error": "Missing subscription fields"})
			return
		}
		if err := h.Store.UpsertPushSubscription(ctx, userID, body.Endpoint, body.Keys.P256dh, body.Keys.Auth); err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
			return
		}
		writeJSON(w, http.StatusOK, map[string]bool{"ok": true})
		return
	}

	if r.Method == http.MethodDelete {
		var body struct {
			Endpoint string `json:"endpoint"`
		}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil || body.Endpoint == "" {
			writeJSON(w, http.StatusBadRequest, map[string]string{"error": "Missing endpoint"})
			return
		}
		_ = h.Store.DeletePushSubscription(ctx, userID, body.Endpoint)
		writeJSON(w, http.StatusOK, map[string]bool{"ok": true})
		return
	}

	writeJSON(w, http.StatusMethodNotAllowed, map[string]string{"error": "Method not allowed"})
}

// HandleNotify handles POST /api/push?action=notify (service key auth).
func (h *PushHandler) HandleNotify(w http.ResponseWriter, r *http.Request) {
	authHeader := r.Header.Get("Authorization")
	if !strings.HasPrefix(authHeader, "Bearer ") {
		writeJSON(w, http.StatusUnauthorized, map[string]string{"error": "Missing authorization"})
		return
	}
	token := strings.TrimPrefix(authHeader, "Bearer ")
	serviceKey := os.Getenv("ZITADEL_SERVICE_USER_PAT")
	if serviceKey == "" || token != serviceKey {
		writeJSON(w, http.StatusUnauthorized, map[string]string{"error": "Invalid authorization"})
		return
	}

	var body struct {
		ForumlineID string `json:"forumline_id"`
		UserID      string `json:"user_id"`
		Title       string `json:"title"`
		Body        string `json:"body"`
		Link        string `json:"link"`
		ForumDomain string `json:"forum_domain"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid request body"})
		return
	}
	if body.Title == "" || body.Body == "" {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "Missing title or body"})
		return
	}

	ctx := r.Context()
	targetUserID := body.UserID
	if targetUserID == "" && body.ForumlineID != "" {
		exists, _ := h.Store.ProfileExists(r.Context(), body.ForumlineID)
		if !exists {
			writeJSON(w, http.StatusNotFound, map[string]string{"error": "User not found"})
			return
		}
		targetUserID = body.ForumlineID
	}
	if targetUserID == "" {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "Missing user_id or forumline_id"})
		return
	}

	// Check mute
	if body.ForumDomain != "" {
		forumID := h.Store.GetForumIDByDomain(ctx, body.ForumDomain)
		if forumID != "" {
			muted, err := h.Store.IsNotificationsMuted(ctx, targetUserID, forumID)
			if err != nil {
				writeJSON(w, http.StatusInternalServerError, map[string]string{"error": "Failed to check mute status"})
				return
			}
			if muted {
				writeJSON(w, http.StatusOK, map[string]interface{}{"ok": true, "skipped": "forum_muted"})
				return
			}
		}
	}

	sent := h.PushService.SendToUser(ctx, targetUserID, body.Title, body.Body, body.Link, body.ForumDomain)
	writeJSON(w, http.StatusOK, map[string]interface{}{"ok": true, "sent": sent})
}
